### PR TITLE
Libreswan 4.00 ikeports

### PR DESCRIPTION
**Pull request by mangelajo**
Related-Issue: #683

### CODE REVIEW

**Comment by mangelajo** [comment]
Still not fully tested

Probably I should remove the true/false and just keep this patch ready for when libreswan is bumped.

**Comment by mangelajo** [comment]
> If we want to avoid tying this to a specific Libreswan package upgrade, we could check the version at runtime: `ipsec --version` outputs a string of the form `Linux Libreswan 3.32 (netkey) on 5.7.7-100.fc31.x86_64`.

Yeah, althought there is some problem with the libreswan logic to generate the packages (at least from git), it generates 3.30-xxxxxx-nnn 

I tried to do auto-detection based on the pluto output, but the exit error in that case is asynchronouseon the go func... and it didn't work.

**Comment by skitt** [comment]
> > If we want to avoid tying this to a specific Libreswan package upgrade, we could check the version at runtime: `ipsec --version` outputs a string of the form `Linux Libreswan 3.32 (netkey) on 5.7.7-100.fc31.x86_64`.
> 
> Yeah, althought there is some problem with the libreswan logic to generate the packages (at least from git), it generates 3.30-xxxxxx-nnn

Ah, yes, the tags aren’t on the main branch after 3.30 :-/.

> I tried to do auto-detection based on the pluto output, but the exit error in that case is asynchronouseon the go func... and it didn't work.

How about `ipsec pluto --help` and `ipsec whack --help`, and looking for `ikeport` in the output (on stderr)?

**Comment by mangelajo** [comment]
> > > If we want to avoid tying this to a specific Libreswan package upgrade, we could check the version at runtime: `ipsec --version` outputs a string of the form `Linux Libreswan 3.32 (netkey) on 5.7.7-100.fc31.x86_64`.
> > 
> > 
> > Yeah, althought there is some problem with the libreswan logic to generate the packages (at least from git), it generates 3.30-xxxxxx-nnn
> 
> Ah, yes, the tags aren’t on the main branch after 3.30 :-/.
> 
> > I tried to do auto-detection based on the pluto output, but the exit error in that case is asynchronouseon the go func... and it didn't work.
> 
> How about `ipsec pluto --help` and `ipsec whack --help`, and looking for `ikeport` in the output (on stderr)?

I think it's simpler if we just wait for the release, although the transition will be less seamless but we will have 2 benefits:
1) Knowing exactly when we got the new libreswan
2) The code is simpler

As you said it's actually easy to do if we look at help.

**Comment by mangelajo** [comment]
@skitt If you think seamless transition is better I can handle it too.. I'm just not sure ':D

**Comment by skitt** [comment]
Test PR in https://github.com/submariner-io/submariner/pull/904

File: pkg/cable/libreswan/libreswan.go [hunk 1]
```diff
@@ -262,13 +262,15 @@ func (i *libreswan) ConnectToEndpoint(endpoint types.SubmarinerEndpoint) (string
 				args = append(args, "--id", localEndpointIdentifier)
 				args = append(args, "--host", localEndpointIP)
 				args = append(args, "--client", leftSubnets[lsi])
+				args = append(args, "--ikeport", i.ipSecNATTPort)
```
Comment (skitt) [inline pkg/cable/libreswan/libreswan.go:265]: Doesn’t this need to be handled differently for NAT and non-NAT?

```suggestion
if endpoint.Spec.NATEnabled {
    args = append(args, "--ikeport", i.ipSecNATTPort)
} else {
    args = append(args, "--ikeport", i.ipSecPort)
}
```

and same again for the right-hand side.

**Review by skitt — CHANGES_REQUESTED** [review]

File: pkg/cable/libreswan/libreswan.go (hunk 1 above)
Comment (mangelajo) [reply pkg/cable/libreswan/libreswan.go:265]: Yes,  that makes sense.

Honestly I guess it doesn't really matter if we always go over a single port and remove the concept of nat/non-nat port ? 

I believe IKEv2 can do both over 4500 , may be all over "ipSecPort" and that's it?

Comment (skitt) [reply pkg/cable/libreswan/libreswan.go:265]: Yup, e2e works over 4500 in non-NAT mode.

**Comment by mangelajo** [comment]
Let's handle the bits here as part of : https://github.com/submariner-io/submariner/pull/904

**Pull request closed without merging**